### PR TITLE
fix: DebugBar block by CSP

### LINKED ISSUE #8405: Bug: [DebugBar] block by CSP
### PHP Version

8.2

### CodeIgniter4 Version

4.4.4

### CodeIgniter4 Installation Method

Composer (using `codeigniter4/appstarter`)

### Which operating systems have you tested for this bug?

Windows, Linux

### Which server did you use?

apache

### Database

_No response_

### What happened?

Debugbar use inline style ,  in mode development and CSPEnabled true will show many error in console browser

Screenshot 2024-01-06 115033

Screenshot 2024-01-06 125259

### Steps to Reproduce

set  CSPEnabled = true, 
set ENVIRONMENT to development

### Expected Output

No error 

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by YapsBridging**
**Description**
Fixes #8405
- Remove inline style : javascrip:void ,onclick,display none
- Update CSS  for show tab 
- Add switchClass func to switch class CSS
- Add eventListener

 **BEFORE**
Screenshot 2024-01-11 202320

 **AFTER**
Screenshot 2024-01-11 110823

**Steps to Reproduce**
set CSPEnabled = true,
set ENVIRONMENT to development

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: admin/css/debug-toolbar/toolbar.scss
Comment (michalsn) [inline admin/css/debug-toolbar/toolbar.scss:479]:
```suggestion
#ci-timeline.debug-bar-dblock {
```

File: admin/css/debug-toolbar/toolbar.scss
Comment (michalsn) [inline admin/css/debug-toolbar/toolbar.scss:528]:
```suggestion
.debug-bar-dtableRow {
    display: table-row;
}
```

File: admin/css/debug-toolbar/toolbar.scss
Comment (michalsn) [inline admin/css/debug-toolbar/toolbar.scss:530]:
```suggestion
.debug-bar-dinlineBlock {
```
The same goes for the rest of the styles, please.

File: admin/css/debug-toolbar/toolbar.scss
Comment (michalsn) [inline admin/css/debug-toolbar/toolbar.scss:542]:
```suggestion
.debug-bar-level-0{
```
Please add a  scope.

File: system/Debug/Toolbar/Views/toolbar.js
Comment (michalsn) [inline system/Debug/Toolbar/Views/toolbar.js:52]: Can't we just make it: `el.classList.replace()`?

File: system/Debug/Toolbar/Views/toolbar.js
Comment (michalsn) [inline system/Debug/Toolbar/Views/toolbar.js:74]:
```suggestion
            if (toggleData === "datatable") {
```
Please fix styling of the code - for the entire code block (mostly spaces)

File: system/Debug/Toolbar/Views/toolbar.js
Comment (michalsn) [inline system/Debug/Toolbar/Views/toolbar.js:162]: Can be replaced with the one-liner: `el.classList.replace()` - not sure if we need this method.

File: system/Debug/Toolbar/Views/toolbar.js
Comment (michalsn) [inline system/Debug/Toolbar/Views/toolbar.js:180]:
```suggestion
            if (target.classList.contains("debug-bar-ndisplay")) {
                target.classList.replace("debug-bar-ndisplay", "debug-bar-dtableRow");
            } else {
                target.classList.replace("debug-bar-dtableRow", "debug-bar-ndisplay");
            } 
```

File: system/Debug/Toolbar/Views/toolbar.js
Comment (michalsn) [inline system/Debug/Toolbar/Views/toolbar.js:199]: The same can be applied here.

File: system/Debug/Toolbar/Views/toolbar.js
Comment (michalsn) [inline system/Debug/Toolbar/Views/toolbar.js:220]: Please fix the styling.

**Review by michalsn — CHANGES_REQUESTED** [review]
Look nice, but please fix the styling of the CSS and JS files.

**Comment by YapsBridging** [comment]
Thank you @michalsn for the explanation,
previously I was confused about what was wrong,...I'll try it

File: system/Debug/Toolbar/Views/toolbar.js
Comment (YapsBridging) [reply system/Debug/Toolbar/Views/toolbar.js:162]: @michalsn , Sometimes the class to be replaced doesn't exist, but we still have to add the new class,
If we use classList.replace()  then this doesn't work

ex.
we must add class **dblock** n remove class **ndisplay**
but in some condition the element do not have yet class **ndisplay**

so if we use replace. the function cant work and the **dblock** class will not be added to the element

Comment (michalsn) [reply system/Debug/Toolbar/Views/toolbar.js:162]: Ok, that's fine. I haven't looked in-depth at what html is generated, so if there is a reason to have the `switchClass` method - go for it.

**Comment by kenjis** [comment]
We don't use merge commits in PR branch. 
If you can use git, please use `git rebase`.
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#updating-your-branch

File: app/Views/welcome_message.php
Comment (kenjis) [inline app/Views/welcome_message.php:334]:
```suggestion
<script <?= csp_script_nonce() ?>>
```

**Comment by kenjis** [comment]
Please sign all commits.
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#gpg-signing-old-commits

**Comment by YapsBridging** [comment]
> We don't use merge commits in PR branch. If you can use git, please use `git rebase`. See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#updating-your-branch

thanks @kenjis , This is my first time making a PR,
sorry if there are any mistakes,
so what can I do to fix this with a rebase? can you tell me the steps?

I have read the link above, but I don't understand how to fix it, can you help me

and sorry my bad english.

File: app/Views/welcome_message.php
Comment (YapsBridging) [reply app/Views/welcome_message.php:334]: thanks @kenjis , but why dont use  {csp-script-nonce} ? 
like line 12 welcome_message.php,  use {csp-style-nonce}

File: system/Debug/Toolbar/Views/toolbar.js
Comment (YapsBridging) [reply system/Debug/Toolbar/Views/toolbar.js]: thanks @michalsn

**Comment by kenjis** [comment]
@YapsBridging Welcome to CodeIgniter4!

First, I recommend you set up your git to sign all commits (in this repository).
> To configure your Git client to sign commits by default for a local repository, in Git versions 2.0.0 and above, run `git config commit.gpgsign true`. To sign all commits by default in any local repository on your computer, run `git config --global commit.gpgsign true`.
https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

See also https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/signing.md

**Comment by kenjis** [comment]
Set upstream repository:
```
git remote add upstream git@github.com:codeigniter4/CodeIgniter4.git
```

Fetch the latest upstream:
```
git fetch upstream
```

Rebase your PR branch:
```
git switch csp
git rebase upstream/develop
```

Force push:
```
git push --force-with-lease origin HEAD
```

File: app/Views/welcome_message.php
Comment (kenjis) [reply app/Views/welcome_message.php:334]: Okay, you are correct. The line 12 uses `<style {csp-style-nonce}>`.
So using it here is okay.

I personally think we don't need to use `{csp-style-nonce}` if we can use the `csp_script_nonce()` function. It is a CI4 global function and we can always use it in PHP scripts.

Comment (YapsBridging) [reply app/Views/welcome_message.php:334]: i think so too..
if csp is enable ,it  must following change the placeholder tagNonce , cause If not change and if an attacker injects a string like <script {csp-script-nonce}> it might become the real nonce attribute with functionality.
https://www.codeigniter.com/user_guide/outgoing/csp.html

so the best we can use csp_script_nonce()

**Comment by YapsBridging** [comment]
> Set upstream repository:
> 
> ```
> git remote add upstream git@github.com:codeigniter4/CodeIgniter4.git
> ```
> 
> Fetch the latest upstream:
> 
> ```
> git fetch upstream
> ```
> 
> Rebase your PR branch:
> 
> ```
> git switch csp
> git rebase upstream/develop
> ```
> 
> Force push:
> 
> ```
> git push --force-with-lease origin HEAD
> ```

thank @kenjis 
I have tried the steps above, and I got a big problem..haha, sorry im a newbe..

so can i create a new PR with new Branch and close/delete this one?

**Comment by kenjis** [comment]
Oh, you don't use git. Creating a new PR is no problem!

But if you could use git, you can change your commits easily without creating new PRs.
If you want to learn git, I would help you.
Or if you help to improve our contribution guide, I would appreciate it.
- https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#how-to-guide
- https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#syncing-your-repository

**Comment by YapsBridging** [comment]
> Oh, you don't use git. Creating a new PR is no problem!
> 
> But if you could use git, you can change your commits easily without creating new PRs. If you want to learn git, I would help you. Or if you help to improve our contribution guide, I would appreciate it.
> 
> * https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#how-to-guide
> * https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#syncing-your-repository

thanks @kenjis , of course i want,

i'm glad ,if u want to help me to,

basically i use git with command prompt but just standard command pull, push ,commit create branch,switch, check log,etc,,

one of the problems is when i try to signing to all  old commit with

git rebase -i --root --exec 'git commit --amend --no-edit --no-verify -S'

https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#gpg-signing-old-commits

its also overwrites  other people's commits with my signature
i dont understand how this work..LOL

maybe because of the merge.

Screenshot 2024-01-25 194106

LoL..
so can I go back to the old commit before I merged, before all these problems happened?

**Comment by kenjis** [comment]
You did not push your local branch to the remote branch in GitHub.
So you still have the remote branch (copy) in your local repository.

```
$ git switch develop
$ git branch -D csp # delete local csp branch
$ git checkout csp
branch 'csp' set up to track 'origin/csp'.
Switched to a new branch 'csp'
```

**Comment by kenjis** [comment]
Or you can clone your repository in GitHub again.
```
git clone git@github.com:YapsBridging/CodeIgniter4.git
git checkout csp
```

**Comment by kenjis** [comment]
I recommend:
> As a faster alternative, you can still securely sign commits without the -S option in git commit by setting git config --global commit.gpgsign true and git config --global user.signingkey 3AC5C34371567BD2 to all local repositories. Without the --global option, the change is applied to one local repository only.
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#gpg-signing-old-commits

and if you set it, `git rebase` also automatically signs all rebased commits.

If you follow the [instruction](https://github.com/codeigniter4/CodeIgniter4/pull/8411#issuecomment-1905118431), the commit log would be following:
```
a7a441716c (HEAD -> csp) update style & js
795de34a07 update style & js
539c8b3c02 fix changes SCSS files did not match the expected CSS output
915b3251f1 remove inline style debugbar,welcome_message for CSP
7c426fa1dd update css for show tab debugbar
219011070a (upstream/develop) Merge pull request #8453 from kenjis/test-postgre-upsertBatch
```

**Comment by YapsBridging** [comment]
@kenjis , Ok I'll try it

**Comment by YapsBridging** [comment]
greate !..

Screenshot 2024-01-25 224501

one step again
i try  to Force push:
git push --force-with-lease origin HEAD , 
and  i get this message..what happen?

Screenshot 2024-01-25 224936

**Comment by kenjis** [comment]
You did `git rebase` successfully!

The error on push seems an issue about your token permission.
See https://stackoverflow.com/questions/64059610/how-to-resolve-refusing-to-allow-an-oauth-app-to-create-or-update-workflow-on

I use SSH auth, so I have never seen that error.
```
$ git remote -v
origin	git@github.com:YapsBridging/CodeIgniter4.git (fetch)
origin	git@github.com:YapsBridging/CodeIgniter4.git (push)
upstream	git@github.com:codeigniter4/CodeIgniter4.git (fetch)
upstream	git@github.com:codeigniter4/CodeIgniter4.git (push)
```

**Comment by YapsBridging** [comment]
hi @kenjis , I have completed the permission token permission, 
i try again
git push --force-with-lease origin HEAD ,

now i get message 
Screenshot 2024-02-02 114209

can u tell me how to fix it ?

**Comment by kenjis** [comment]
It shows someone (probably you?) added commit(s) that is not included in your local branch to the remote branch on GitHub.

If you are sure the local branch is no problem, try:
```
# Update your local branch, because the upstream develop advanced.
git fetch upstream
git switch csp
git rebase upstream/develop
# Force push with no check.
git push -f origin HEAD
```

**Comment by kenjis** [comment]
@YapsBridging Can't you push to your PR branch?

**Comment by YapsBridging** [comment]
> @YapsBridging Can't you push to your PR branch?

@kenjis , I thought I had completed the token permission. but apparently not, I need some more time to push again..., maybe with a new branch, I'm also afraid of lots of conflicts when I push using the same branch

**Comment by kenjis** [comment]
The steps in https://github.com/codeigniter4/CodeIgniter4/pull/8411#issuecomment-1923119478 did not work?

Or pushing as a new branch and creating a new PR would be fine.

**Comment by YapsBridging** [comment]
> The steps in [#8411 (comment)](https://github.com/codeigniter4/CodeIgniter4/pull/8411#issuecomment-1923119478) did not work?
> 
> Or pushing as a new branch and creating a new PR would be fine.

I haven't tried it yet,
I'm afraid that when I push there will be lots of conflicts, I have already committed to the remote branch, and also the merge that I did is already on the remote branch. CSP's remote branch contains merge and commit, and my local branch doesn't

**Comment by kenjis** [comment]
I don't understand the conflicts you are afraid of.
*Force push* pushes your local branch as it is. It can remove the commits in the remote branch on GitHub.

**Comment by YapsBridging** [comment]
> I don't understand the conflicts you are afraid of. _Force push_ pushes your local branch as it is. It can remove the commits in the remote branch on GitHub.

ow, i dont know  about that,,,thanks @kenjis , i will try push again

File: admin/css/debug-toolbar/toolbar.scss
Comment (paulbalandan) [inline admin/css/debug-toolbar/toolbar.scss:478]: I believe this can be simplified to one CSS class, `.debug-bar-dblock`, instead of individually showing all uses in `#ci-*` ids. If there would be a new id, say `#ci-dump`, it would need to be added here again.

File: app/Views/welcome_message.php
Comment (kenjis) [inline app/Views/welcome_message.php:153]: What is `f`?

File: admin/css/debug-toolbar/toolbar.scss
Comment (YapsBridging) [reply admin/css/debug-toolbar/toolbar.scss:478]: You're right..thanks @paulbalandan

File: app/Views/welcome_message.php
Comment (YapsBridging) [reply app/Views/welcome_message.php:153]: @kenjis  its font

Comment (kenjis) [reply app/Views/welcome_message.php:153]: It seems we don't use names like `f1`.
Can you change to meaningful names?
`font-something`?

Comment (YapsBridging) [reply app/Views/welcome_message.php:153]: ok @kenjis  i'll change it

**Comment by kenjis** [comment]
macOS and Firefox.

Before:
Screenshot 2024-02-12 14 53 36

After:
Screenshot 2024-02-12 14 54 00

**Review by michalsn — APPROVED** [review]
LGTM!

**Review by paulbalandan — APPROVED** [review]

**Review by kenjis — APPROVED** [review]
Thank you!

**Comment by crustamet** [comment]
This is great, but there is one little issue if only one case exist... while having 

for example
`<base href="localhost:8080">`
in your main html the buttons on debugbar always redirect to base if you click the buttons "Database", "Views", "Logs", "Files"

I always get redirected to "homepage" using this new debugbar. i think this happened before but without the CSP Fix while clicking on them nothing happened...

The fix is as following remove the href="#" from the button links

`

	<a href="#" data-toggle="datatable" data-table="<?= strtolower(str_replace(' ', '-', $heading)) ?>">
		<h2><?= $heading ?></h2>
	</a>

	TO

	<a data-toggle="datatable" data-table="<?= strtolower(str_replace(' ', '-', $heading)) ?>">
		<h2><?= $heading ?></h2>
	</a>

`

**Comment by kenjis** [comment]
What do you mean?

It works with no problem with the following settings.
```diff
diff --git a/app/Config/App.php b/app/Config/App.php
index 21b4df2052..aa433bf847 100644
--- a/app/Config/App.php
+++ b/app/Config/App.php
@@ -174,5 +174,5 @@ class App extends BaseConfig
      * @see http://www.html5rocks.com/en/tutorials/security/content-security-policy/
      * @see http://www.w3.org/TR/CSP/
      */
-    public bool $CSPEnabled = false;
+    public bool $CSPEnabled = true;
 }
diff --git a/app/Views/welcome_message.php b/app/Views/welcome_message.php
index 86a6e39de3..88001613ca 100644
--- a/app/Views/welcome_message.php
+++ b/app/Views/welcome_message.php
@@ -2,6 +2,7 @@
 <html lang="en">
 <head>
     <meta charset="UTF-8">
+    <base href="localhost:8080">
     <title>Welcome to CodeIgniter 4!</title>
     <meta name="description" content="The small framework with powerful features">
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
```

**Comment by YapsBridging** [comment]
@crustamet  i see, cause tag base add value to All relative link in the page
@kenjis  this will happen if the url  page different from the tag base 

ex in 
head tag base http://localhost:8080
but page url http://localhost:8080/dashboard

so the tag/button debugbar will be http://localhost:8080/#  

before i change javascript:void(0) to #  , because javascript:void(0) was blocked by CSP

**Comment by YapsBridging** [comment]
but in my opinion, if you use tag base, maybe other plugins will also be affected. If you use other plugins, you don't know whether the plugin you are using contains relative links.

the tag Base  will add value to All  relative link in the page

So in my opinion using the base tag is not appropriate, it can cause unexpected problems.

that's just my opinion

**Comment by crustamet** [comment]
> but in my opinion, if you use tag base, maybe other plugins will also be affected. If you use other plugins, you don't know whether the plugin you are using contains relative links.
> the tag Base will add value to All relative link in the page
> **So in my opinion using the base tag is not appropriate, it can cause unexpected problems**.
> that's just my opinion

It only causes problems if you have already create the project that is not using this **base** option and then you use it when the project is finished :)) it doesnt work like that, you use it from the beginning of the project creation so it helps you organize everything better

I don't think this matters for is appropriate or not, it does not break any functionalities for me, and its better because i load all my links with this base link.

And i don't think it matters in this situation either, because what ? i added a valid base meta tag and the debugbar is not working ? :))) the debugbar should work whatever you do with your code.

**Comment by YapsBridging** [comment]
ok  @crustamet,  you're right, maybe this is the reason why the debugbar button  uses javascript:void(0) not #

**Comment by crustamet** [comment]
why in the world would you want to have a hashtag in your links without wanting them there ?
the hashtag was before used in the old times to remember what you opened..

Here in the new debugbar it does that by adding data-(*attrs) to the links so now there is no need for href="#" at this point
It should be removed completely from the links

It would be better to have them as buttons instead of links but yeah there is bug with `<base>` tag the debugbar not working..
It only works on the homepage of the project

**Comment by crustamet** [comment]
you should really make them with buttons 😆  but **a** tags without **href** so weird

**Comment by ScorsIndonesia** [comment]
ow i forgot ...Thank you  @kenjis , for teaching Git...:))

**Comment by YapsBridging** [comment]
ow i forgot...Thank you @kenjis , for teaching Git....:))

**Comment by kenjis** [comment]
> @kenjis  this will happen if the url  page different from the tag base 
>
> ex in 
head tag base http://localhost:8080
but page url http://localhost:8080/dashboard
>
> so the tag/button debugbar will be http://localhost:8080/#  

Cannot reproduce with `<base href="localhost:8080">` on Firefox.
The tag/button debugbar is `http://localhost:8080/dashboard#`

**Comment by kenjis** [comment]
Oh, I was able to reproduce with `<base href="http://localhost:8080">` and `http://localhost:8080/dashboard`.

**Pull request merged**
